addons/controllers/kapp-controller/kappcontrollerconfig_controller.go

```diff
@@ -0,0 +1,258 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	"github.com/go-logr/logr"
+
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
+	runv1alpha3 "github.com/vmware-tanzu/tanzu-framework/apis/run/v1alpha3"
+)
+
+const KappControllerAddonName = "kapp-controller"
+
+// KappControllerConfigReconciler reconciles a KappControllerConfig object
+type KappControllerConfigReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=run.tanzu.vmware.com,resources=kappcontrollerconfigs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=run.tanzu.vmware.com,resources=kappcontrollerconfigs/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *KappControllerConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("kappControllerConfig", req.NamespacedName)
+
+	// get kapp-controller config object
+	kappControllerConfig := &runv1alpha3.KappControllerConfig{}
+	if err := r.Client.Get(ctx, req.NamespacedName, kappControllerConfig); err != nil {
+		log.Error(err, "unable to fetch kappControllerConfig")
+		return ctrl.Result{}, err
+	}
+
+	// get the parent cluster name from owner reference
+	// if the owner reference doesn't exist, use the same name as config CRD
+	clusterNamespacedName := req.NamespacedName
+	cluster := &clusterapiv1beta1.Cluster{}
+	for _, owner := range kappControllerConfig.OwnerReferences {
+		if owner.Kind == cluster.Kind {
+			clusterNamespacedName.Name = owner.Name
+			break
+		}
+	}
+
+	// verify that the cluster related to config is present
+	if err := r.Client.Get(ctx, clusterNamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Cluster not found")
+			return ctrl.Result{}, nil
+		}
+
+		log.Error(err, "unable to fetch cluster")
+		return ctrl.Result{}, err
+	}
+
+	if retResult, err := r.ReconcileKappControllerConfig(ctx, log, cluster, kappControllerConfig); err != nil {
+		log.Error(err, "unable to reconcile kappControllerConfig")
+		return retResult, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KappControllerConfigReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&runv1alpha3.KappControllerConfig{}).
+		WithOptions(options).
+		Complete(r)
+}
+
+// ReconcileKappControllerConfig reconciles KappControllerConfig CRD
+func (r *KappControllerConfigReconciler) ReconcileKappControllerConfig(
+	ctx context.Context,
+	log logr.Logger,
+	cluster *clusterapiv1beta1.Cluster,
+	kappControllerConfig *runv1alpha3.KappControllerConfig) (_ ctrl.Result, retErr error) {
+
+	var (
+		patchConfig bool
```

Comment:
After some thoughts, I think it might be better to keep this `patchConfig ` bool. We explicitly set it to be true when there are diffs, instead of relying the patch function to calculate the diff, will make our intent more clear. @danniel1205

Reply:
Why though? The point of `clusterapipatchutil.NewHelper` is to check for diff and patch if needed.

Reply:
Removed `patchConfig `